File: packages/sdk-node/src/instruments/anthropic.ts
import { randomUUID } from 'node:crypto';
import { AgentBeamClient } from '../client';
import { SpanData } from '../types';

/**
 * Wrap an Anthropic client so every `messages.create` call is automatically
 * traced and reported to AgentBeam.
 *
 * Uses `Proxy` so the returned value has the exact same type as the input --
 * callers retain full type safety and autocomplete.
 */
export function wrapAnthropic<T>(client: T, ab: AgentBeamClient): T {
  const raw = client as Record<string, unknown>;
  const originalMessages = raw.messages as Record<string, unknown>;
  const originalCreate = originalMessages.create as (...args: unknown[]) => unknown;

  // Proxy the `messages` namespace so we can intercept `create`
  const messagesProxy = new Proxy(originalMessages, {
    get(target, prop, receiver) {
      if (prop === 'create') {
        return function wrappedCreate(...args: unknown[]) {
          const params = (args[0] ?? {}) as Record<string, unknown>;
          const startedAt = new Date();
          const traceId = randomUUID();
          const spanId = randomUUID();

          // Build input preview from the messages array
          const inputMessages = params.messages as
            | Array<{ role: string; content: unknown }>
            | undefined;
          let inputPreview = '';
          if (inputMessages && Array.isArray(inputMessages)) {
            const last = inputMessages[inputMessages.length - 1];
            if (last) {
              inputPreview =
                typeof last.content === 'string'
                  ? last.content.slice(0, 4096)
                  : JSON.stringify(last.content).slice(0, 4096);
            }
          }

          const modelName = (params.model as string) ?? 'unknown';

          // Call the real method
          const result = originalCreate.apply(target, args);

          // Handle both Promise and non-Promise returns
          if (result && typeof (result as { then?: unknown }).then === 'function') {
            return (result as Promise<unknown>).then(
              (response) => {
                recordSpan(ab, response, {
                  traceId,
                  spanId,
                  startedAt,
                  modelName,
                  inputPreview,
                });
                return response;
              },
              (err: unknown) => {
                recordErrorSpan(ab, err, {
                  traceId,
                  spanId,
                  startedAt,
                  modelName,
                  inputPreview,
                });
                throw err;
              },
            );
          }

          // Synchronous return (unlikely but safe to handle)
          recordSpan(ab, result, {
            traceId,
            spanId,
            startedAt,
            modelName,
            inputPreview,
          });
          return result;
        };
      }
      return Reflect.get(target, prop, receiver);
    },
  });

  // Proxy the top-level client to return our patched `messages`
  return new Proxy(client as object, {
    get(target, prop, receiver) {
      if (prop === 'messages') return messagesProxy;
      return Reflect.get(target, prop, receiver);
    },
  }) as T;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

interface SpanContext {
  traceId: string;
  spanId: string;
  startedAt: Date;
  modelName: string;
  inputPreview: string;
}

function recordSpan(
  ab: AgentBeamClient,
  response: unknown,
  ctx: SpanContext,
): void {
  const endedAt = new Date();
  const res = response as Record<string, unknown> | null | undefined;

  const usage = res?.usage as
    | { input_tokens?: number; output_tokens?: number }
    | undefined;

  // Extract output preview from the first content block
  let outputPreview = '';
  const content = res?.content as Array<{ type: string; text?: string }> | undefined;
  if (content && Array.isArray(content) && content.length > 0) {
    const textBlock = content.find((b) => b.type === 'text');
    if (textBlock?.text) {
      outputPreview = textBlock.text.slice(0, 4096);
    }
  }

  const inputTokens = usage?.input_tokens ?? 0;
  const outputTokens = usage?.output_tokens ?? 0;

  const span: SpanData = {
    trace_id: ctx.traceId,
    span_id: ctx.spanId,
    span_name: 'anthropic.messages.create',
    span_kind: 'llm',
    status: 'ok',
    model_provider: 'anthropic',
    model_name: (res?.model as string) ?? ctx.modelName,
    input_tokens: inputTokens,
    output_tokens: outputTokens,
    total_tokens: inputTokens + outputTokens,
    started_at: ctx.startedAt.toISOString(),
    ended_at: endedAt.toISOString(),
    duration_ms: endedAt.getTime() - ctx.startedAt.getTime(),
    input_preview: ctx.inputPreview,
    output_preview: outputPreview,
  };

  ab.addSpan(span);
}

function recordErrorSpan(
  ab: AgentBeamClient,
  err: unknown,
  ctx: SpanContext,
): void {
  const endedAt = new Date();

  const errorMessage =
    err instanceof Error ? err.message : String(err);
  const errorType =
    err instanceof Error ? err.constructor.name : 'UnknownError';

  const span: SpanData = {
    trace_id: ctx.traceId,
    span_id: ctx.spanId,
    span_name: 'anthropic.messages.create',
    span_kind: 'llm',
    status: 'error',
    model_provider: 'anthropic',
    model_name: ctx.modelName,
    started_at: ctx.startedAt.toISOString(),
    ended_at: endedAt.toISOString(),
    duration_ms: endedAt.getTime() - ctx.startedAt.getTime(),
    input_preview: ctx.inputPreview,
    error_message: errorMessage,
    error_type: errorType,
  };

  ab.addSpan(span);
}

File: packages/sdk-python/agentbeam/instruments/anthropic.py
from __future__ import annotations

import time
import uuid
from datetime import datetime, timezone
from typing import Any, TYPE_CHECKING

if TYPE_CHECKING:
    from agentbeam.client import AgentBeamClient


def patch_anthropic(client: AgentBeamClient) -> None:
    """Monkey-patch the Anthropic SDK to auto-trace all messages.create calls."""
    try:
        import anthropic
        import anthropic.resources
    except ImportError:
        return

    original_create = anthropic.resources.Messages.create

    def patched_create(self: Any, *args: Any, **kwargs: Any) -> Any:
        from agentbeam.types import SpanData

        trace_id = str(uuid.uuid4())
        span_id = str(uuid.uuid4())
        start_time = time.time()
        started_at = datetime.now(timezone.utc).isoformat()

        model = kwargs.get("model", args[0] if args else "unknown")
        messages = kwargs.get("messages", [])

        # Build input preview from the last message
        input_preview = ""
        if messages:
            last_msg = messages[-1]
            content = last_msg.get("content", "")
            if isinstance(content, list):
                text_parts = [
                    p.get("text", "")
                    for p in content
                    if isinstance(p, dict) and p.get("type") == "text"
                ]
                content = " ".join(text_parts)
            input_preview = str(content)[:500]

        try:
            response = original_create(self, *args, **kwargs)
            duration = int((time.time() - start_time) * 1000)

            # Extract token usage
            usage = getattr(response, "usage", None)
            input_tokens = getattr(usage, "input_tokens", 0) if usage else 0
            output_tokens = getattr(usage, "output_tokens", 0) if usage else 0

            # Extract output preview
            output_preview = ""
            content_blocks = getattr(response, "content", [])
            if content_blocks:
                text_blocks = [
                    b.text for b in content_blocks if hasattr(b, "text")
                ]
                output_preview = " ".join(text_blocks)[:500]

            span: SpanData = {
                "trace_id": trace_id,
                "span_id": span_id,
                "span_name": "anthropic.messages.create",
                "span_kind": "llm",
                "status": "ok",
                "model_provider": "anthropic",
                "model_name": model,
                "input_tokens": input_tokens,
                "output_tokens": output_tokens,
                "total_tokens": input_tokens + output_tokens,
                "started_at": started_at,
                "ended_at": datetime.now(timezone.utc).isoformat(),
                "duration_ms": duration,
                "input_preview": input_preview,
                "output_preview": output_preview,
            }
            client.record_span(span)
            return response

        except Exception as e:
            duration = int((time.time() - start_time) * 1000)

            span: SpanData = {  # type: ignore[no-redef]
                "trace_id": trace_id,
                "span_id": span_id,
                "span_name": "anthropic.messages.create",
                "span_kind": "llm",
                "status": "error",
                "model_provider": "anthropic",
                "model_name": model,
                "started_at": started_at,
                "ended_at": datetime.now(timezone.utc).isoformat(),
                "duration_ms": duration,
                "error_message": str(e),
                "error_type": type(e).__name__,
            }
            client.record_span(span)
            raise

    anthropic.resources.Messages.create = patched_create  # type: ignore[assignment]
